### PR TITLE
[C-API] Implement tensor_if custom condition and add test cases

### PR DESCRIPTION
**Pull request by gichan-jang**
Implement tensor_if custom register/unregister and add related test cases.

Self evaluation:
1. Build test: [ *]Passed [ ]Failed [ ]Skipped
2. Run test: [ *]Passed [ ]Failed [ ]Skipped

### CODE REVIEW

File: api/capi/src/nnstreamer-capi-pipeline.c [hunk 1]
```diff
@@ -2159,3 +2160,119 @@ ml_pipeline_custom_easy_filter_unregister (ml_custom_easy_filter_h custom)
...
+  int status = 0;
+  guint i;
+  ml_if_custom_s *c;
+  ml_tensors_data_h in_data;
+  ml_tensors_data_s *_data;
+  ml_tensors_info_h ml_info;
+  GstTensorsInfo in_info = *info;
+  gboolean ret = FALSE;
+
+
```
Comment (dongju-chae) [inline api/capi/src/nnstreamer-capi-pipeline.c:2180]: Double lines

File: api/capi/src/nnstreamer-capi-pipeline.c [hunk 2]
```diff
@@ -2159,3 +2160,119 @@ ml_pipeline_custom_easy_filter_unregister (ml_custom_easy_filter_h custom)
...
+  g_free (custom->name);
+  g_free (custom);
+}
+
+/**
+ * @brief Registers the tensor_if custom callback.
+ */
+int ml_pipeline_tensor_if_custom_register (const char *name, ml_pipeline_if_custom_cb cb, void *user_data, ml_pipeline_if_h *if_custom)
+{
+    int status = ML_ERROR_NONE;
```
Comment (dongju-chae) [inline api/capi/src/nnstreamer-capi-pipeline.c:2224]: indent?

File: api/capi/src/nnstreamer-capi-pipeline.c [hunk 3]
```diff
@@ -2159,3 +2160,119 @@ ml_pipeline_custom_easy_filter_unregister (ml_custom_easy_filter_h custom)
...
+  check_feature_state ();
+
+  if (!name || !cb || !if_custom)
+    return ML_ERROR_INVALID_PARAMETER;
+
+  /* init null */
+  *if_custom = NULL;
+
+  /* create and init custom handle */
+  if ((c = g_new0 (ml_if_custom_s, 1)) == NULL)
```
Comment (dongju-chae) [inline api/capi/src/nnstreamer-capi-pipeline.c:2236]: `g_new0` cannot be failed.

File: api/capi/src/nnstreamer-capi-pipeline.c [hunk 4]
```diff
@@ -2159,3 +2160,119 @@ ml_pipeline_custom_easy_filter_unregister (ml_custom_easy_filter_h custom)
...
+  check_feature_state ();
+
+  if (!if_custom)
+    return ML_ERROR_INVALID_PARAMETER;
+
+  c = (ml_if_custom_s *) if_custom;
+
+  if (nnstreamer_if_custom_unregister (c->name) != 0) {
+    ml_loge ("Failed to unregister tensor_if custom condition %s.", c->name);
+    return ML_ERROR_INVALID_PARAMETER;
```
Comment (dongju-chae) [inline api/capi/src/nnstreamer-capi-pipeline.c:2273]: Is it really `ML_ERROR_INVALID_PARAMETER`?

File: api/capi/include/nnstreamer.h [hunk 5]
```diff
@@ -925,7 +925,40 @@ int ml_pipeline_element_get_property_enum (ml_pipeline_element_h elem_h, const c
  *
  * Here is an example of the usage:
  * @code
- * TBU
+ * // Define callback for tensor_if custom condition.
+ * static int tensor_if_custom_cb (const ml_tensors_info_h info, const ml_tensors_data_h data, void *user_data, int *result)
+ * {
+ *  // Describe the conditions and pass the results
```
Comment (jaeyun-jung) [inline api/capi/include/nnstreamer.h:931]: it is better to add comment of result value (result 1 and 0 for true/false case, return 0; if no error)

File: api/capi/src/nnstreamer-capi-pipeline.c (hunk 3 above)
Comment (jaeyun-jung) [reply api/capi/src/nnstreamer-capi-pipeline.c:2236]: use g_try_new0 istead

File: tests/tizen_capi/unittest_tizen_capi.cc [hunk 6]
```diff
@@ -10000,6 +10000,208 @@ TEST (nnstreamer_capi_custom, register_filter_10_n)
...
+/**
+ * @brief Test for tensor_if custom registration.
+ * @detail Invalid params.
+ */
+TEST (nnstreamer_capi_if, register_03_n)
+{
+  int status;
+
+  /* test code with null param */
+  status = ml_pipeline_tensor_if_custom_register ("tifx", test_if_custom_cb, NULL, NULL);
```
Comment (jaeyun-jung) [inline tests/tizen_capi/unittest_tizen_capi.cc:10188]: lets cover another cases in this function (name or cb is null)

File: api/capi/include/nnstreamer.h [hunk 7]
```diff
@@ -925,7 +925,40 @@ int ml_pipeline_element_get_property_enum (ml_pipeline_element_h elem_h, const c
  *
  * Here is an example of the usage:
  * @code
- * TBU
+ * // Define callback for tensor_if custom condition.
+ * static int tensor_if_custom_cb (const ml_tensors_info_h info, const ml_tensors_data_h data, void *user_data, int *result)
```
Comment (jaeyun-jung) [inline api/capi/include/nnstreamer.h:929]: plz do rebase and fix params.

File: tests/tizen_capi/unittest_tizen_capi.cc [hunk 8]
```diff
@@ -10000,6 +10000,208 @@ TEST (nnstreamer_capi_custom, register_filter_10_n)
...
+/**
+ * @brief Test for tensor_if custom condition
+ */
+TEST (nnstreamer_capi_if, custom_01_p)
+{
+  const gchar *_tmpdir = g_get_tmp_dir ();
+  const gchar *_dirname = "nns-tizen-XXXXXX";
+  gchar *fullpath = g_build_path ("/", _tmpdir, _dirname, NULL);
+  gchar *dir = g_mkdtemp ((gchar *)fullpath);
+  gchar *file = g_build_path ("/", dir, "output", NULL);
```
Comment (jaeyun-jung) [inline tests/tizen_capi/unittest_tizen_capi.cc:10036]: need free string file and dir

File: api/capi/include/nnstreamer.h (hunk 7 above)
Comment (gichan-jang) [reply api/capi/include/nnstreamer.h]: Rebased and fixed params. Thanks!

File: api/capi/include/nnstreamer.h (hunk 5 above)
Comment (gichan-jang) [reply api/capi/include/nnstreamer.h]: Added. Thanks.

File: api/capi/src/nnstreamer-capi-pipeline.c (hunk 1 above)
Comment (gichan-jang) [reply api/capi/src/nnstreamer-capi-pipeline.c]: Updated!

File: api/capi/src/nnstreamer-capi-pipeline.c (hunk 2 above)
Comment (gichan-jang) [reply api/capi/src/nnstreamer-capi-pipeline.c]: Updated!

File: api/capi/src/nnstreamer-capi-pipeline.c (hunk 3 above)
Comment (gichan-jang) [reply api/capi/src/nnstreamer-capi-pipeline.c]: Changed to g_try_new0.

File: api/capi/src/nnstreamer-capi-pipeline.c (hunk 4 above)
Comment (gichan-jang) [reply api/capi/src/nnstreamer-capi-pipeline.c]: Thanks. Changed to `ML_ERROR_STREAMS_PIPE`.

File: tests/tizen_capi/unittest_tizen_capi.cc (hunk 8 above)
Comment (gichan-jang) [reply tests/tizen_capi/unittest_tizen_capi.cc:10036]: Oh.. Thanks. Updated

File: tests/tizen_capi/unittest_tizen_capi.cc (hunk 6 above)
Comment (gichan-jang) [reply tests/tizen_capi/unittest_tizen_capi.cc:10190]: resolved.

**Review by dongju-chae — APPROVED** [review]
LGTM

**Comment by gichan-jang** [comment]
The parameter order of `ml_pipeline_if_custom_cb` has been changed to maintain consistency with the existing callbacks.

File: api/capi/src/nnstreamer-capi-pipeline.c [hunk 9]
```diff
@@ -2159,3 +2160,119 @@ ml_pipeline_custom_easy_filter_unregister (ml_custom_easy_filter_h custom)
...
+
+  /* init null */
+  *if_custom = NULL;
+
+  /* create and init custom handle */
+  if ((c = g_try_new0 (ml_if_custom_s, 1)) == NULL)
+    return ML_ERROR_OUT_OF_MEMORY;
+
+  c->name = g_strdup (name);
```
Comment (wooksong) [inline api/capi/src/nnstreamer-capi-pipeline.c:2238]: Where is ```c->name``` freed?

Comment (gichan-jang) [reply api/capi/src/nnstreamer-capi-pipeline.c:2240]: free from `ml_pipeline_if_custom_free`.

**Review by jaeyun-jung — APPROVED** [review]

**Review by myungjoo — APPROVED** [review]

**Pull request merged**